### PR TITLE
NAS-115546 / 22.12 / Refactor ManualUpdateComponent

### PR DESCRIPTION
**Pull request by RehanY147**

### CODE REVIEW

File: src/app/pages/system/general-settings/support/file-ticket-form/file-ticket-form.component.ts [hunk 1]
```diff
@@ -215,7 +215,7 @@ export class FileTicketFormComponent implements OnInit {
   openSuccessDialog(params: NewTicketResponse): void {
     const conf: DialogFormConfiguration = {
       title: this.translate.instant('Ticket'),
-      message: this.translate.instant('Congratulations! Your ticket has been submitted successfully. It may take some time before images appear.'),
+      message: this.translate.instant('Your ticket has been submitted successfully. It may take some time before images appear.'),
```
Comment (undsoft) [inline src/app/pages/system/general-settings/support/file-ticket-form/file-ticket-form.component.ts:218]: Didn't like the happy tone? :)

File: src/app/pages/system/update/manual-update-form/manual-update-form.component.ts [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+import { MessageService } from 'app/modules/entity/entity-form/services/message.service';
+import { EntityJobComponent } from 'app/modules/entity/entity-job/entity-job.component';
+import { EntityUtils } from 'app/modules/entity/utils';
+import { DialogService, SystemGeneralService, WebSocketService } from 'app/services';
+import { AppState } from 'app/store';
+import { waitForSystemInfo } from 'app/store/system-info/system-info.selectors';
+
+@UntilDestroy()
+@Component({
+  selector: 'manual-update-form',
```
Comment (undsoft) [inline src/app/pages/system/update/manual-update-form/manual-update-form.component.ts:33]: Unnecessary selector.

File: src/app/pages/system/update/manual-update-form/manual-update-form.component.ts [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+  templateUrl: './manual-update-form.component.html',
+  providers: [MessageService],
+  styleUrls: ['manual-update-form.component.scss'],
+})
+export class ManualUpdateFormComponent implements OnInit {
+  isFormLoading$: Subject<boolean> = new BehaviorSubject(false);
```
Comment (undsoft) [inline src/app/pages/system/update/manual-update-form/manual-update-form.component.ts:39]: First type is no necessary.
What's the reason to use BehaviorSubject?

File: src/app/pages/system/update/manual-update-form/manual-update-form.component.ts [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+import { AppState } from 'app/store';
+import { waitForSystemInfo } from 'app/store/system-info/system-info.selectors';
+
+@UntilDestroy()
+@Component({
+  selector: 'manual-update-form',
+  templateUrl: './manual-update-form.component.html',
+  providers: [MessageService],
+  styleUrls: ['manual-update-form.component.scss'],
+})
```
Comment (undsoft) [inline src/app/pages/system/update/manual-update-form/manual-update-form.component.ts:37]: ChangeDetection.OnPush

File: src/app/pages/system/update/manual-update-form/manual-update-form.component.ts [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+export class ManualUpdateFormComponent implements OnInit {
+  isFormLoading$: Subject<boolean> = new BehaviorSubject(false);
+  form = this.formBuilder.group({
+    filelocation: ['', Validators.required],
+    updateFile: [null as FileList],
+    rebootAfterManualUpdate: [false],
+  });
+  updateFile = '';
```
Comment (undsoft) [inline src/app/pages/system/update/manual-update-form/manual-update-form.component.ts:45]: unused

File: src/app/pages/system/update/manual-update-form/manual-update-form.component.ts [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+  private get apiEndPoint(): string {
+    return '/_upload?auth_token=' + this.ws.token;
+  }
+
+  userPrefs: Preferences = null;
```
Comment (undsoft) [inline src/app/pages/system/update/manual-update-form/manual-update-form.component.ts:50]: You can use `selectPreferences` to get user preferences. To update them, you'd need to create an action in `preferences.actions.ts`.

File: src/app/pages/system/update/manual-update-form/manual-update-form.component.ts [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    private store$: Store<AppState>,
+  ) { }
+
+  ngOnInit(): void {
+    this.checkHaLicenseAndUpdateStatus();
+    this.getVersionNoFromSysInfo();
+    this.setPoolOptions();
+    this.getUserPrefs();
+    this.mapTempFileLocationValueToNull();
+  }
```
Comment (undsoft) [inline src/app/pages/system/update/manual-update-form/manual-update-form.component.ts:75]: 👍🏻

File: src/app/pages/system/update/manual-update-form/manual-update-form.component.ts [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+      this.currentVersion = sysInfo.version;
+    });
+  }
+
+  setPoolOptions(): void {
+    this.ws.call('pool.query').pipe(untilDestroyed(this)).subscribe((pools) => {
+      if (!pools) {
+        return;
+      }
+      const options = [{ label: 'Memory device', value: ':temp:' }];
```
Comment (undsoft) [inline src/app/pages/system/update/manual-update-form/manual-update-form.component.ts:112]: Missing translation.

File: src/app/pages/system/update/manual-update-form/manual-update-form.component.html [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+          [label]="helptext.filename.placeholder | translate"
+          [tooltip]="helptext.filename.tooltip | translate"
+          [multiple]="false"
+          [required]="true"
+          [acceptedFiles]="'.tar,.update'"
+        ></ix-file-input>
+        <ix-checkbox
+          formControlName="rebootAfterManualUpdate"
+          [label]="helptext.rebootAfterManualUpdate.placeholder"
+          [tooltip]="helptext.rebootAfterManualUpdate.tooltip"
```
Comment (undsoft) [inline src/app/pages/system/update/manual-update-form/manual-update-form.component.html:28]: | translate

File: src/app/pages/system/update/manual-update-form/manual-update-form.component.html [hunk 10]
```diff
@@ -0,0 +1,45 @@
+<mat-card class="card">
+  <mat-card-content>
+    <mat-progress-bar *ngIf="isFormLoading$ | async" mode="indeterminate"></mat-progress-bar>
+    <form [formGroup]="form" class="ix-form-container" (submit)="onSubmit()">
+      <ix-fieldset [title]="'General Options' | translate">
```
Comment (undsoft) [inline src/app/pages/system/update/manual-update-form/manual-update-form.component.html:5]: Title doesn't really do much.

File: src/app/pages/system/update/manual-update-form/manual-update-form.component.html [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+          [tooltip]="helptext.filelocation.tooltip | translate"
+          [options]="fileLocationOptions$"
+          [required]="true"
+        ></ix-select>
+
+        <ix-file-input
+          formControlName="updateFile"
+          [label]="helptext.filename.placeholder | translate"
+          [tooltip]="helptext.filename.tooltip | translate"
+          [multiple]="false"
```
Comment (undsoft) [inline src/app/pages/system/update/manual-update-form/manual-update-form.component.html:21]: Looks like multiple false is broken, as I can select multiple files.

File: src/app/pages/system/update/manual-update-form/manual-update-form.component.ts [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+        options.push({
+          label: '/mnt/' + pool.name, value: '/mnt/' + pool.name,
+        });
+      });
+      this.fileLocationOptions$ = of(options);
+    });
+  }
+
+  get isEnterprise(): boolean {
+    return this.window.localStorage.getItem('product_type').includes(ProductType.Enterprise);
```
Comment (undsoft) [inline src/app/pages/system/update/manual-update-form/manual-update-form.component.ts:123]: Best to check for equality with ProductType.ScaleEnterprise.

File: src/app/pages/system/update/manual-update-form/manual-update-form.component.html [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        
+      </ix-fieldset>
+     
+      <div class="form-actions">
+        <button
+          mat-button
+          type="submit"
+          [disabled]="form.invalid || (isFormLoading$ | async)"
+          color="primary"
+        >{{ 'Save' | translate }}</button>
```
Comment (undsoft) [inline src/app/pages/system/update/manual-update-form/manual-update-form.component.html:40]: Apply Update

**Review by undsoft — CHANGES_REQUESTED** [review]
On my local machine I get 401 Unauthorized when I submit the form.

File: src/app/pages/system/general-settings/support/file-ticket-form/file-ticket-form.component.ts (hunk 1 above)
Comment (RehanY147) [reply src/app/pages/system/general-settings/support/file-ticket-form/file-ticket-form.component.ts:218]: It was only a ticket. Nothing to call home about :P

**Comment by RehanY147** [comment]
@undsoft I can't reproduce the unauthorized issue. Can you show me the steps please?

**Review by undsoft — CHANGES_REQUESTED** [review]
I get the error when I select Memory device in the form. Then request fails  net::ERR_CONNECTION_RESET.

Also if you need to save user preferences, you likely need to add your new action to preferences.effects.ts

**Comment by RehanY147** [comment]
Merging with master is causing issues. Replacing this PR with https://github.com/truenas/webui/pull/6709

**Pull request closed without merging**